src/core/reactor_backend.cc

```diff
@@ -1143,10 +1143,21 @@ bool reactor_backend_selector::has_enough_aio_nr() {
     auto aio_nr = read_first_line_as<unsigned>("/proc/sys/fs/aio-nr");
     /* reactor_backend_selector::available() will be execute in early stage,
      * it's before io_setup() issued, and not per-cpu basis.
+     *
+     * As smp::count is initialized with 0 in early stages, set local_count
+     * to 1. If/when this method is called after smp::count has been properly
+     * initialized, then its actual value will be used.
+     *
      * So this method calculates:
      *  Available AIO on the system - (request AIO per-cpu * ncpus)
      */
-    if (aio_max_nr - aio_nr < reactor::max_aio * smp::count) {
+    unsigned smp_local_count = 1;
+
+    if (smp::count > 0){
```

Comment:
This check is never true. The `available()` method (the only caller of _this_ one) is only called eary when options are initialized (and when smp::count is not set). I guess this function had lost it chances to work correctly after the CLI options rework.

Probably the best we can do at this point is -- always initialize CLI with full list of available backends, but the `reactor_backend_selector::create` should check `has_enough_aio_nr() && detect_aio_poll()` before trying to create the AIO one. And, probably, throw if it cannot.

Reply:
Makes sense. I submitted a new change which always initialize with the full list of backend, and - within `reactor_backend_selector::create` added the `has_enough_aio_nr() && detect_aio_poll()` check. 

Whenever the default backend fails, it will print a warning and fallback to epoll. Similarly, when epoll is explicitly requested by the user, we will go straight to it.